### PR TITLE
refactor(gnoland): add subcommands support

### PR DESCRIPTION
**Pull request by moul**
BREAKING CHANGE: CLI usage changed for node operators.

* [x] add subommands support
* [x] rename `gnoland` to `gnoland server`
* [x] update doc
* [x] update deployment scripts
* [x] fix tests, CI/CD
* [x] ~add a `gnoland init` command~ (won't do now, we need to choose between keeping lazy init in `server` + having `init` VS having always to run the two commands because `server` can't init)

Addresses #731 

<details><summary>Checklists...</summary>

## Contributors Checklist

- [x] Added new tests, or not needed, or not feasible
- [x] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [x] Updated the official documentation or not needed
- [x] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [x] Added references to related issues and PRs
- [x] Provided any useful hints for running manual tests
- [x] Added new benchmarks to [generated graphs](https://gnoland.github.io/benchmarks), if any. More info [here](../.benchmarks/README.md).

## Maintainers Checklist

- [ ] Checked that the author followed the guidelines in `CONTRIBUTING.md`
- [ ] Checked the conventional-commit (especially PR title and verb, presence of `BREAKING CHANGE:` in the body)
- [ ] Ensured that this PR is not a significant change or confirmed that the review/consideration process was appropriate for the change
</details>

### CODE REVIEW

**Review by thehowl — COMMENTED** [review]
Code looks OK. What other subcommand would you like to add which would justify having subcommands? (init, in my opinion, could just be a flag...). Mostly asking as I don't think it makes sense to add the need to specify the "server" subcommand, both because it breaks old documentation and because it seems redundant (seeing as the gnoland binary's main "thing" is the server)

**Comment by moul** [comment]
Current status:

- `gnoland --init-only` is now `gnoland init`
- `gnoland` is now `gnoland server`

---

@albttx explained that it's common to have those subcommands so that managing a node becomes more flexible.
See https://github.com/gnolang/gno/issues/731#issuecomment-1592754593

Here is what we have when running a common cosmos-sdk chain (note that it mixes node management, client and other things):

```
Available Commands:
  add-genesis-account    Add a genesis account to genesis.json
  collect-gentxs         Collect genesis txs and output a genesis.json file
  config                 Create or query an application CLI configuration file
  debug                  Tool for helping with debugging your application
  export                 Export state to JSON
  export-derive-balances Export a derive balances from a provided genesis export
  forceprune             Forceprune option prunes and compacts blockstore.db and state.db.
  gentx                  Generate a genesis tx carrying a self delegation
  get-env                Prints the current environment
  help                   Help about any command
  init                   Initialize private validator, p2p, genesis, and application configuration files
  keys                   Manage your application's keys
  migrate                Migrate genesis to a specified target version
  prepare-genesis        Prepare a genesis file with initial setup
  query                  Querying subcommands
  rollback               rollback tendermint state by one height
  rosetta                spin up a rosetta server
  set-env                Set home environment variables for commands
  staked-to-csv          Export a airdrop csv from a provided balances export
  start                  Run the full node
  status                 Query remote node for status
  tendermint             Tendermint subcommands
  testnet                Initialize files for a simapp testnet
  tx                     Transactions subcommands
  unsafe-reset-all       (unsafe) Remove all the data and WAL, reset this node's validator to genesis state
  validate-genesis       validates the genesis file at the default location or at the location passed as an arg
  version                Print the application binary version information
```

It would be beneficial to consider adding additional commands such as gnoland doctor or utility functions for working with the store, such as extracting contracts or state.

I will rename `server` to `start`, which looks more standard.

File: gnovm/cmd/gno/test_test.go [hunk 1]
```diff
@@ -186,5 +186,7 @@ func TestTest(t *testing.T) {
 		// {args: []string{"test", "../../tests/integ/failing2", "--verbose"}, stderrShouldBe: "=== PREC  ./../../tests/integ/failing2\n=== BUILD ./../../tests/integ/failing2\n=== RUN   file/failing_filetest.gno\n", recoverShouldBe: "fail on ../../tests/integ/failing2/failing_filetest.gno: got unexpected error: beep boop"},
 		// {args: []string{"test", "../../../examples/gno.land/p/demo/ufmt", "--verbose", "--timeout", "10000" /* 10µs */}, recoverShouldContain: "test timed out after"}, // FIXME: should be testable
 	}
-	testMainCaseRun(t, tc)
+	for i := 0; i < 100; i++ {
+		testMainCaseRun(t, tc)
+	}
```
Comment (thehowl) [inline gnovm/cmd/gno/test_test.go:191]: I imagine you added this to provide the pprof visualizations you shared with us. Do you want to keep this?

Maybe have this as a different test which is skipped with `t.Short()`?

File: go.mod
Comment (thehowl) [inline go.mod:1]: I imagine dep changes also coming from visualizations?

**Review by thehowl — APPROVED** [review]

File: gno.land/cmd/gnoland/root.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+		_, _ = fmt.Fprintf(os.Stderr, "%+v\n", err)
+		os.Exit(1)
+	}
+}
+
+func newRootCmd(io *commands.IO) *commands.Command {
+	cmd := commands.NewCommand(
+		commands.Metadata{
+			ShortUsage: "<subcommand> [flags] [<arg>...]",
+			LongHelp:   "Starts the gnoland blockchain node",
```
Comment (harry-hov) [inline gno.land/cmd/gnoland/root.go:27]: Maybe?
```suggestion
			ShortHelp:  "Starts the gnoland blockchain node",
```

File: gno.land/cmd/gnoland/start.go [hunk 3]
```diff
@@ -41,32 +37,23 @@ type gnolandCfg struct {
...
 		commands.Metadata{
-			ShortUsage: "[flags] [<arg>...]",
-			LongHelp:   "Starts the gnoland blockchain node",
-			Options: []ff.Option{
-				ff.WithConfigFileFlag("config"),
-				ff.WithConfigFileParser(fftoml.Parser),
-			},
+			Name:       "start",
+			ShortUsage: "start [flags]",
+			ShortHelp:  "run the full node",
```
Comment (harry-hov) [inline gno.land/cmd/gnoland/start.go:47]: Capitalize the first letter?
```suggestion
			ShortHelp:  "Run the full node",
```

**Review by harry-hov — APPROVED** [review]

**Comment by harry-hov** [comment]
Maybe we also need to modify this[1]

```diff
-./build/gnoland
+./build/gnoland start
```
[1] https://github.com/moul/gno/blob/46baeaa8518760dbb56e2b11f96aa6f8a8595b81/gno.land/cmd/gnofaucet/README.md?plain=1#L19

File: gnovm/cmd/gno/test_test.go (hunk 1 above)
Comment (moul) [reply gnovm/cmd/gno/test_test.go:191]: Oops, yes shouldn't have been committed. Will revert.

File: go.mod
Comment (moul) [reply go.mod:1]: what do you mean?

File: gnovm/cmd/gno/test_test.go (hunk 1 above)
Comment (moul) [reply gnovm/cmd/gno/test_test.go:191]:
```suggestion
	testMainCaseRun(t, tc)
```

**Comment by dongwon8247** [comment]
Changed the [Developer Portal](https://docs.onbloc.xyz/docs/environment-setup/build-the-local-testnet#step-1.-run-a-local-node) docs accordingly

**Pull request merged**